### PR TITLE
Introduce a ResetPulseBridge to Remove An Early Deadlock Condition

### PR DESCRIPTION
**Pull request by davidbiancolin**
Currently, reset in the target is implemented using the peek-poke widget, via a blocking call to `simif_t::target_reset`. Since other bridges are not being ticked while this function executes, the simulator can deadlock near time zero if any bridge stalls waiting on an invocation of `tick` (e.g., a synthesized assertion fires). 

This PR introduces a dedicated bridge that can produce a runtime-configurable reset pulse to drive the target. This will subsume target_reset(), allowing the simulator to enter it's main loop immediately upon simulation commence. 

Before i can remove target_reset from firesim_top, i'll need to instantiate the bridge in Chipyard. I will open separate PRs to do that after the release has been handled, and this PR has been merged. 

This is based off the hostIF PR (#778); i'll retarget to dev once it is merged.

### CODE REVIEW

File: sim/midas/src/main/cc/bridges/reset_pulse.cc
Comment (timsnyder-siv) [inline sim/midas/src/main/cc/bridges/reset_pulse.cc:1]: What is this `RESETPULSEBRIDGEMODULE_struct_guard` macro and how does it get defined?  I'm not seeing that in this pr, is it just letting you have this code in your branch but not use it in the driver right now?  Is this supposed to be taken out before merging?

Comment (davidbiancolin) [reply sim/midas/src/main/cc/bridges/reset_pulse.cc:1]: This is a necessary (evil) component of defining a bridge since one of the constructor arguments is defined for the first time in the generated header. If no bridges are instantiated, the struct is not defined, and this code cannot compile...

**Review by timsnyder-siv — APPROVED** [review]

File: sim/src/main/cc/firesim/firesim_top.cc
Comment (davidbiancolin) [reply sim/src/main/cc/firesim/firesim_top.cc:569]: I'm going to hold off on removing these two lines until we've bumped CY.

**Pull request merged**
